node/directives/auditAccess.ts

```diff
@@ -0,0 +1,64 @@
+import type { GraphQLField } from 'graphql'
+import { defaultFieldResolver } from 'graphql'
+import { SchemaDirectiveVisitor } from 'graphql-tools'
+
+import sendAuthMetric, { AuthMetric } from '../metrics/auth'
+import { checkUserPermission } from '../resolvers/Queries/Users'
+
+export class AuditAccess extends SchemaDirectiveVisitor {
+  public visitFieldDefinition(field: GraphQLField<any, any>) {
+    const { resolve = defaultFieldResolver } = field
+
+    field.resolve = async (
+      root: any,
+      args: any,
+      context: Context,
+      info: any
+    ) => {
+      const {
+        vtex: { adminUserAuthToken, storeUserAuthToken, account, logger },
+        request,
+      } = context
+
+      const operation = field.astNode?.name?.value ?? request.url
+      const forwardedHost = request.headers['x-forwarded-host'] as string
+      const caller = request.headers['x-vtex-caller'] as string
+
+      const hasAdminToken = !!(
+        adminUserAuthToken ?? (context?.headers.vtexidclientautcookie as string)
+      )
+
+      const hasStoreToken = !!storeUserAuthToken
+      const hasApiToken = !!request.headers['vtex-api-apptoken']
+
+      let role
+      let permissions = []
+
+      if (hasAdminToken || hasStoreToken) {
+        const checkUserPermissions = await checkUserPermission(
```

Comment:
It would be best for this resolver not to wait on any part of the data fetching for sending metrics, so that the response can be returned faster (as well as avoiding an error response due to this).

The `sendMetric` call further below is already happening on the background, but not this one. A simple way to make sure the response doesn't wait for any part of this logic is to extract this whole function as an async function defined outside `visitFieldDefinition`. For example:

```js
async function sendAuthMetric(field: GraphQLField, context: Context) {
  // Evertyhing from the original function except for the `resolve()` call.
  // You can await as usual within this function, even on the `sendMetric` call.
}
```

Then you can call this `sendAuthMetric` function without `await. That way the resolver won't wait for any part of it, truly running all async parts on the background. Something like this:

```js
    field.resolve = async (
      root: any,
      args: any,
      context: Context,
      info: any
    ) => {
      sendAuthMetric(field, context)
      return resolve(root, args, context, info)
    }
```

Let me know if that makes sense.